spmd/tensor/ops/matrix_ops.py

```diff
@@ -1,27 +1,60 @@
 # Copyright (c) Meta Platforms, Inc. and affiliates
 # implement matrix related ops for distributed tensor
-from typing import Optional
-from spmd.tensor.dispatch import OpSchema
-from spmd.tensor.placement_types import PlacementSpec
+from spmd.tensor.dispatch import OpSchema, OutputSharding
 from spmd.tensor.ops.prop_rules import einop_prop, mm_prop, pointwise_prop
+from spmd.tensor.placement_types import _Partial, PlacementSpec, Shard
 from spmd.tensor.ops.utils import register_prop_rule
 
 
 @register_prop_rule("aten.mm.default")
-def mm_rules(op_schema: OpSchema) -> Optional[PlacementSpec]:
+def mm_rules(op_schema: OpSchema) -> OutputSharding:
     mat1_spec, mat2_spec = op_schema.args_spec
-    return mm_prop(mat1_spec, mat2_spec)
+    return OutputSharding(mm_prop(mat1_spec, mat2_spec))
 
 
 @register_prop_rule("aten.addmm.default")
-def addmm_rules(op_schema: OpSchema) -> Optional[PlacementSpec]:
+def addmm_rules(op_schema: OpSchema) -> OutputSharding:
     input_spec, mat1_spec, mat2_spec = op_schema.args_spec
     mm_out_spec = mm_prop(mat1_spec, mat2_spec)
     if mm_out_spec is None:
-        return None
-    return pointwise_prop((input_spec, mm_out_spec))
+        # non-eligible input, suggest addmm input specs
+        # TODO: add suggested input specs for resharding
+        return OutputSharding(None)
+    # TODO: add multi dim support for addmm
+    if (
+        mm_out_spec.placements[0].is_partial()
+        and input_spec.placements[0].is_replicate()
+    ):
+        # for the case where mm out is partial and input is replicate
+        # we should suggest the input be a partial instead of replicate
+        suggested_input_spec = PlacementSpec(
+            input_spec.ndim, input_spec.mesh, [_Partial()]
+        )
+        suggested_mat1_spec = PlacementSpec(
+            mat1_spec.ndim, mat1_spec.mesh, [Shard(1)]
+        )
+        suggested_mat2_spec = PlacementSpec(
+            mat2_spec.ndim, mat2_spec.mesh, [Shard(0)]
+        )
+        return OutputSharding(
+            None,
+            schema_suggestions=[
+                OpSchema(
+                    (
+                        suggested_input_spec,
+                        suggested_mat1_spec,
+                        suggested_mat2_spec,
+                    ),
+                    op_schema.kwargs_schema,
+                )
+            ],
+        )
+
+    return OutputSharding(
+        pointwise_prop((input_spec, mm_out_spec), linear=True)
+    )
```

Comment:
What I meant is to replace this with the more general pointwis_rules implementation. THis is like calling add(Partial, Replicated), so it generalizes to a pointwise operation. The pointwise sharding propagation should take this input and then return a suggestion of Partial, Partial --- then we'd back-propagate the suggestion to the output.  (it's like chaining suggestions). This will only work if the pointwise rule reutrns a suggestion that keeps the "partial" input untouched though but we could make it in a way that ensures that's the case.

Reply:
The reason for doing this is to reuse the code that will support multiple mesh dims.

Reply:
Oh i see, got it, yeah this make sense, will update